bundles/org.eclipse.passage.loc.licenses.core/src/org/eclipse/passage/loc/internal/licenses/core/FloatingLicensePackFromRequest.java

```diff
@@ -187,33 +189,39 @@ private Collection<FeatureGrant> featureGrants(FloatingLicensePack pack) {
 
 	private FeatureGrant featureGrant(LicensePlanFeatureDescriptor feature, FloatingLicensePack pack, int no) {
 		FeatureGrant grant = new EmptyFeatureGrant().get();
-		String fid = feature.getFeature().getIdentifier();
-		grant.getFeature().setIdentifier(fid);
+		grant.getFeature().setIdentifier(feature.getFeature().getIdentifier());
 		grant.setCapacity(request.defaultCapacity());
 		grant.setIdentifier(String.format("%s#%d", request.identifier(), no)); //$NON-NLS-1$
 		grant.setPack(pack);
-		grant.setValid(featureGrantPeriod(fid));
-		grant.setVivid(featureGrantVivid(fid));
+		grant.setValid(featureGrantPeriod(feature.getFeature()));
+		grant.setVivid(featureGrantVivid(feature.getFeature()));
 		grant.getFeature().setVersionMatch(version(feature));
 		return grant;
 	}
 
-	private ValidityPeriod featureGrantPeriod(String feature) {
+	private ValidityPeriod featureGrantPeriod(FeatureRefDescriptor feature) {
 		return template//
 				.flatMap(l -> forFeature(l.getFeatures(), feature)) //
+				.filter(g -> g.eIsSet(LicensesPackage.eINSTANCE.getFeatureGrant_Vivid())) //
```

Comment:
I have mention something in my initial comment, but I don't know it exactly anymore. I can just say that the call of grant.setVivid() can have issues then, but a missing vivid should probably be handled at a higher level.

> If `forFeature()` not always returns an empty Optional `featureGrantPeriod()` should check if the vivid is set because `getVivid()` returns a long primitive which is never boxed to null. Filtering out Grants whose vivid is not set ensures that the given default (60) is used instead of zero in that case.